aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+  // NB: So, I am pretty sure that get_parameters() does not rely in any way
+  // on the dropout descriptor.  So we fake up a dummy one instead of try
+  // to actually make one legitimately.
+  RNNDescriptor rnn_desc = rnn.descriptor(handle);
+
+  // TODO: allocation here is goofy
+  TensorDescriptor x_desc(any_param.type().tensor({1, input_size}), 5);
+
+  auto num_weights = get_num_weights(handle, rnn_desc, x_desc, rnn.datatype);
+  auto weight_buf = any_param.type().tensor(num_weights).zero_();
+
+  FilterDescriptor w_desc;
+  w_desc.set(weight_buf, 3);
+
+  // Slice off views into weight_buf
+  std::vector<Tensor> params_arr;
+  size_t params_stride0;
+  std::tie(params_arr, params_stride0) = get_parameters(handle, rnn, rnn_desc, x_desc, w_desc, weight_buf);
+
+  MatrixRef<Tensor> weight{weight_arr, static_cast<size_t>(weight_stride0)},
+                    params{params_arr, params_stride0};
+
+  // Copy weights
+  _copyParams(weight, params);
+
+  // Update the storage
+  for (size_t i = 0; i < weight.size(0); i++) {
+    for (auto orig_param_it = weight[i].begin(), new_param_it = params[i].begin();
+         orig_param_it != weight[i].end() && new_param_it != params[i].end();
+         orig_param_it++, new_param_it++) {
+      auto orig_param = *orig_param_it, new_param = *new_param_it;
+      orig_param.set_(new_param.view_as(orig_param));
+    }
+  }
+
+  return weight_buf;
+}
+
+// NB: when fn_batch_sizes is empty, that means no batch sizes was specified
+std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor> _cudnn_rnn(
+    const Tensor& input_r,
+    TensorList weight, int64_t weight_stride0,
+    const Tensor& weight_buf_r, const Tensor& hx, const Tensor& cx,
+    int64_t fn_mode, int64_t fn_hidden_size,
+    int64_t fn_num_layers, bool batch_first, double fn_dropout,
+    bool fn_train, bool fn_bidirectional, IntList fn_batch_sizes,
+    const Tensor& fn_dropout_state
+    ) {
+
+  auto input = input_r;
+  auto weight_buf = weight_buf_r;
+
+  RNNParams fn;
+  fn.rnn.set_mode(fn_mode);
+  fn.rnn.hidden_size = fn_hidden_size;
+  fn.rnn.num_layers = fn_num_layers;
+  fn.rnn.set_bidirectional(fn_bidirectional);
+  fn.rnn.datatype = getCudnnDataType(input);
+  fn.dropout.set(fn_train, fn_dropout, fn_dropout_state);
+  fn.tensors.set(input.sizes(), fn_batch_sizes, batch_first);
+
+  // TODO: Set device to input
+
+  if (fn.rnn.mode != CUDNN_LSTM) {
+    if (cx.defined()) {
+      throw std::runtime_error("rnn: illegal defined cx for non-LSTM RNN");
+    }
+  }
+
+  // TODO: can batch_first be a wrapper around this function?
+  auto is_input_packed = fn.tensors.batch_sizes.size() != 0;
+  if (batch_first && !is_input_packed) {
+    input = input.transpose(0, 1);
+  }
+
+  auto hidden_size = _hidden_size(fn.rnn, fn.tensors);
+  auto output_size = _output_size(fn.rnn, fn.tensors);
+
+  AT_ASSERT(hx.is_contiguous(), "hx.is_contiguous()");
+  AT_ASSERT(!cx.defined() || cx.is_contiguous(), "!cx or cx.is_contiguous()");
```

Comment:
1. I thought we decided on using `*_EXPECT` macros for validating user inputs, because they aren't really assertions
2. You know that these error messages are user facing, right? Can you please make them nicer?

Reply:
They used to be asserts Python side and I didn't think too hard when porting them. Time to fix >:)